library/ccm.c

```diff
@@ -422,27 +424,32 @@ int mbedtls_ccm_self_test( int verbose )
         if( verbose != 0 )
             mbedtls_printf( "  CCM-AES #%u: ", (unsigned int) i + 1 );
 
+        memset( plaintext, 0, PLAINTEXT_MAX_LENGTH );
+        memset( cipertext, 0, CIPHERTEXT_MAX_LENGTH );
+        memcpy( plaintext, msg, msg_len[i]);
+
         ret = mbedtls_ccm_encrypt_and_tag( &ctx, msg_len[i],
-                                   iv, iv_len[i], ad, add_len[i],
-                                   msg, out,
-                                   out + msg_len[i], tag_len[i] );
+                                           iv, iv_len[i], ad, add_len[i],
+                                           plaintext, cipertext,
+                                           cipertext + msg_len[i], tag_len[i] );
 
         if( ret != 0 ||
-            memcmp( out, res[i], msg_len[i] + tag_len[i] ) != 0 )
+            memcmp( cipertext, res[i], msg_len[i] + tag_len[i] ) != 0 )
         {
             if( verbose != 0 )
                 mbedtls_printf( "failed\n" );
 
             return( 1 );
         }
+        memset( plaintext, 0, PLAINTEXT_MAX_LENGTH );
```

Comment:
We're zero'ising the plaintext after use, in a test where we're about to decrypt the same data back into the same buffer. I think this should use `mbedtls_zeroise`.